### PR TITLE
build: initial support for poetry build tool

### PR DESCRIPTION
**Pull request by MatriQ**
# Description

This PR is just adding poetry configurations in `pyproject.toml` so that we can use poety to easily manage the environment, as well as a little of README.md.

Fixes # (issue)

## Type of Change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update, included: [Dify Document](https://github.com/langgenius/dify-docs)
- [x] Improvement, including but not limited to code refactoring, performance optimization, and UI/UX improvement
- [ ] Dependency upgrade

# How Has This Been Tested?

This change doesn't bring out any changes of code, and I have tested with generic poetry commands.

- [ ] TODO

# Suggested Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] I ran `dev/reformat`(backend) and `cd web && npx lint-staged`(frontend) to appease the lint gods
- [ ] `optional` I have made corresponding changes to the documentation 
- [ ] `optional` I have added tests that prove my fix is effective or that my feature works
- [ ] `optional` New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by bowenliang123** [comment]
How to guarantee the `[tool.poetry.dependencies]` settings align to `requireiments.txt`? As well as the `poetry` lock file to the  `requireiments.txt`?

**Comment by MatriQ** [comment]
> How to guarantee the `[tool.poetry.dependencies]` settings align to `requireiments.txt`? As well as the `poetry.lock` lock file to the `requireiments.txt`?

Hi @bowenliang123 , in current stage. we can use `poetry add` command to keep dependencies consistency as I mentioned in README.md: 
image


In the future, I'd like to suggest that remove `requirements.txt` and `requirements-dev.txt` and entirely use poetry to manage dependencies.

**Comment by bowenliang123** [comment]
Can you figure out at least one way in CI jobs to making sure `poetry` dependency list satisfying the existed `requirements.txt` ?

I'm against setting the removal the `requirements.txt` as final target, considering the following facts:
1. `requirements.txt` is generally used out of box with python and pip
2. `poetry` is just one of the build systems with dependencies management
3.  For the runtime usage, the build system does not bring any benefits , while it causes reversed dependencies on `poetry` to the runtime base, like in docker images.

File: api/.python-version [hunk 1]
```diff
@@ -0,0 +1 @@
+3.10
```
Comment (bowenliang123) [inline api/.python-version]: This file is unnecessary, as the `pyporject.toml` has set the required minimum python version. 
Also, dify API module is tested on Python `3.10` and `3.11` in CI jobs.
If this is for local development only, please do not submit to the commit.

**Comment by bowenliang123** [comment]
> This PR is just for providing another for developers to manage dependencies, it doesn't impact unit tests, CI and product env. As for your opsitions, I'd like to discuss with you in another PR(not created yet).

The newly introduced support for the build system should least respect and align to the existing core mechanisms. Otherwise, it will puzzle the users with the issues caused by inconsistencies in dependencies(which happen often).

File: api/.python-version (hunk 1 above)
Comment (MatriQ) [reply api/.python-version]: I'm facing the following issue with 3.11, hence I thought Dify doesn't support 3.11:
image

However, I will remove it.

Comment (bowenliang123) [reply api/.python-version]: The problem you listed is not caused by python 3.11 , I think.
And it should be fixed in #4517, bumping `yfinance` on minor verions.
Please rebase to the latest commit and check again.

Comment (bowenliang123) [reply api/.python-version]: This is an perfect example for inconsistency of your poetry dependencies and the required dependencies from `requirements.txt`.

**Comment by MatriQ** [comment]
> > This PR is just for providing another for developers to manage dependencies, it doesn't impact unit tests, CI and product env. As for your opsitions, I'd like to discuss with you in another PR(not created yet).
> 
> The introduced support for a new build system should at least respect and align with the existing core mechanisms. Otherwise, it will puzzle the users with the issues caused by inconsistencies in dependencies(which happen often), as the de-facto two different dependencies are listed in the Dify API module.

Currently, the only supported build system is still not changed. the dependencies in `pyproject.toml` will be kept update from `requirements.txt`. So there won't introduce any new issues from this PR.

**Comment by bowenliang123** [comment]
> Hi @bowenliang123, is it possible to migrate to poetry in the future? Actually, currently lint and test are using poetry's configuration file, while package dependencies are using pip. I hope in the future we can consolidate everything with poetry, and poetry can parallelize the dependency fetching process to speed up package installation.

Hi @takatost , I am always open to modern build system like hatch or poetry. All I want to prevent is the inconsistency of dependencies.

File: api/.python-version (hunk 1 above)
Comment (MatriQ) [reply api/.python-version]: removed

**Comment by loganclark360** [comment]
?

[W. Logan C - Chat @ Spike](https://spikenow.com/r/a/?ref=spike-organic-signature&_ts=2lrqze)	[2lrqze]

On May 20, 2024 at 14:22 GMT, W Logan C ***@***.***> wrote:

Support

[W. Logan C - Chat @ Spike](https://spikenow.com/r/a/?ref=spike-organic-signature&_ts=2lrqyj)	[2lrqyj]

On May 20, 2024 at 13:52 GMT, langgenius/dify ***@***.***> wrote:

Hi ***@***.***(https://github.com/bowenliang123), is it possible to migrate to poetry in the future? Actually, currently lint and test are using poetry's configuration file, while package dependencies are using pip. I hope in the future we can consolidate everything with poetry, and poetry can parallelize the dependency fetching process to speed up package installation.

Hi ***@***.***(https://github.com/takatost) , I am always open to modern build system like hatch or poetry. All I want to prevent is the inconsistency of dependencies.

**Comment by loganclark360** [comment]
Langgenius is also GitHub!

Thank you

[W. Logan C - Chat @ Spike](https://spikenow.com/r/a/?ref=spike-organic-signature&_ts=2lrv1y)	[2lrv1y]

File: api/.python-version (hunk 1 above)
Comment (bowenliang123) [reply api/.python-version]: Thanks.

**Comment by bowenliang123** [comment]
I'm suggesting the following changes in this PR to meet the baseline of acceptance for initial support for `poetry` build system.

1. change the PR title `build: initial support for poetry build tool`
2. add CI job to check `poetry.lock` file aligned to the dependency list of `[tool.poetry.dependencies]` in `pyproject.toml`, by using `poetry check`
3. add CI job in `api-tests.yml` workflow to install dependencies via poetry and run all through all the tests

These actions still don't guarantee the poetry dependencies aligned to `requirements.txt`, but it should be good enough for initial support. The dependency list of `[tool.poetry.dependencies]` in `pyproject.toml` requires more polishment in version declaration like preferring `^` operator rather than `>=` + `<`, which could be done in the future.

WDYT? @takatost @MatriQ

File: api/pyproject.toml [hunk 2]
```diff
@@ -59,3 +59,108 @@ MOCK_SWITCH = "true"
...
+
+[tool.poetry]
+name = "api"
+version = "0.1.0"
+description = ""
+authors = ["Dify <no-reply@langgenius.com>"]
+readme = "README.md"
+
+[tool.poetry.dependencies]
+python = "^3.10"
```
Comment (bowenliang123) [inline api/pyproject.toml:90]:
```suggestion
python = ">=3.10,<3.12"
```
The api module has been tested on 3.10 and 3.11, but was known to break on 3.12 as for now.

**Comment by MatriQ** [comment]
> I'm suggesting the following changes in this PR to meet the baseline of acceptance for initial support for `poetry` build system.
> 
> 1. change the PR title `build: initial support for poetry build tool`
> 2. add CI job to check `poetry.lock` file aligned to the dependency list of `[tool.poetry.dependencies]` in `pyproject.toml`, by using `poetry check`
> 3. add CI job in `api-tests.yml` workflow to install dependencies via poetry and run all through all the tests
> 
> These actions still don't guarantee the poetry dependencies aligned to `requirements.txt`, but it should be good enough for initial support. The dependency list of `[tool.poetry.dependencies]` in `pyproject.toml` requires more polishment in version declaration like preferring `^` operator rather than `>=` + `<`, which could be done in the future.
> 
> WDYT? @takatost @MatriQ

It's ok for me. But I'd like to get opinion from @takatost

**Review by bowenliang123 — APPROVED** [review]
LGTM overall. It covers the tests with dependencies installed by `peotry`. cc @takatost 
Minor optional suggestions:
1. use `poetry` style version specification like `^` rather than `>=` + `<` for readability and maintainability
2. use `actions-poetry` v3
3. basic info of [tool.poetry] should be updated as
  - name: dify-api
  - version: 0.6.8
  - authors: Dify with its official email

**Review by takatost — DISMISSED** [review]
LGTM

File: api/pyproject.toml [hunk 3]
```diff
@@ -66,3 +66,109 @@ MOCK_SWITCH = "true"
 CODE_MAX_STRING_LENGTH = "80000"
 CODE_EXECUTION_ENDPOINT="http://127.0.0.1:8194"
 CODE_EXECUTION_API_KEY="dify-sandbox"
+
+
+[tool.poetry]
+name = "dify-api"
+version = "0.6.8"
```
Comment (bowenliang123) [inline api/pyproject.toml]:
```suggestion
version = "0.6.10"
```

File: api/pyproject.toml [hunk 4]
```diff
@@ -66,3 +66,109 @@ MOCK_SWITCH = "true"
...
+qrcode = "^7.4.2"
+azure-storage-blob = "12.9.0"
+azure-identity = "1.15.0"
+lxml = "5.1.0"
+xlrd = "^2.0.1"
+pydantic = "^1.10.0"
+pgvecto-rs = "0.1.4"
+firecrawl-py = "0.0.5"
+oss2 = "2.15.0"
+pgvector = "0.2.5"
```
Comment (bowenliang123) [inline api/pyproject.toml:176]: Please update the list by appending recently added dependencies in `requirements.txt` like
```
pymysql
tidb-vector
google-cloud-aiplatform
vanna[postgres,mysql,clickhouse,duckdb]

File: api/pyproject.toml [hunk 5]
```diff
@@ -66,3 +66,109 @@ MOCK_SWITCH = "true"
...
+
+[tool.poetry.group.dev.dependencies]
+coverage = "^7.2.4"
+pytest = "^8.1.1"
+pytest-benchmark = "^4.0.0"
+pytest-env = "^1.1.3"
+pytest-mock = "^3.14.0"
+jinja2 = "^3.1.2"
```
Comment (bowenliang123) [inline api/pyproject.toml]: jinjia2 has been removed from development dependencies.

File: api/pyproject.toml (hunk 3 above)
Comment (takatost) [reply api/pyproject.toml]: Yep. It's best to update to the latest dependencies.

File: api/pyproject.toml (hunk 5 above)
Comment (MatriQ) [reply api/pyproject.toml]: removed

File: api/pyproject.toml (hunk 4 above)
Comment (MatriQ) [reply api/pyproject.toml:176]: Fixed

File: api/pyproject.toml (hunk 3 above)
Comment (MatriQ) [reply api/pyproject.toml]: fixed

**Review by takatost — APPROVED** [review]

File: api/pyproject.toml (hunk 3 above)
Comment (bowenliang123) [reply api/pyproject.toml]: Thanks.

**Pull request merged**